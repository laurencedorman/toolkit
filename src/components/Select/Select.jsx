/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './Select.module.scss';

const Select = ({
  className,
  options,
  placeholder,
  helper,
  error,
  label,
  disabled,
  required,
  themeLight,
  value,
  dataQa,
  ...selectProps
}) => {
  const classNames = cn(
    styles.select,
    styles[className],
    {
      [styles.noOptionSelected]: !value && placeholder,
      [styles.light]: themeLight,
      [styles.hasError]: error,
      [styles.disabled]: disabled,
    },
  );

  return (
    <div className={classNames} data-qa={dataQa}>
      <label>
        {label &&
          <span className={styles.label}>
            {label}{required && <sup>*</sup>}
          </span>
        }
        <span className={styles.mask}>
          <select
            {...selectProps}
            aria-required={required}
            className={styles.control}
            disabled={disabled}
            required={required}
            defaultValue={value}
          >
            {placeholder &&
              <option
                disabled
                value=""
              >
                {placeholder}
              </option>
            }
            {options.map((option, index) => (
              <option
                value={option.value}
                key={`${index}-${option.value}`}
                disabled={option.disabled}
              >
                {option.label}
              </option>
            ))}
          </select>
          <span className={styles.arrow} />
        </span>
      </label>
      {helper && <span className={cn(styles.message, styles.helperMessage)}>{helper}</span>}
      {error && <span className={cn(styles.message, styles.errorMessage)}>{error}</span>}
    </div>
  )
};


Select.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  helper: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  options:  PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string,
    disabled: PropTypes.bool,
  })).isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  themeLight: PropTypes.bool,
  value: PropTypes.string.isRequired,
  dataQa: PropTypes.string,
};

Select.defaultProps = {
  className: undefined,
  disabled: false,
  error: '',
  helper: '',
  label: '',
  placeholder: '',
  required: false,
  themeLight: false,
  dataQa: undefined,
};

export default Select;
