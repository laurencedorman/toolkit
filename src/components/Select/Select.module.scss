@import '~styles/import.scss';

$tinySpace: calc(#{$space} / 4);

.container {
  position: relative;
  width: 100%;

  &:not(:last-child) {
    margin-bottom: $space
  }

  .helper, .error {
    display: block;
    width: 100%;
    margin: 0;
    color: $greyDarker;
    font-size: 0.75rem;
    text-align: left
  }

  .error { color: $cocoCorail }
}

.content {
  .select {
    position: relative;
    display: block;
    width: 100%;
    height: calc(#{$space} * 4);
    background-color: $white;
    background-clip: padding-box;
    box-sizing: border-box;
    font: 1rem/1.618 'Open Sans';

    & > div:first-child {
      width: 100%;
      height: 64px;
      padding: 0;
      border: $border $borderColor;
      border-radius: $radius;
    }

    &.hasError {
      border-color: $cocoCorail
    }
  }
}
