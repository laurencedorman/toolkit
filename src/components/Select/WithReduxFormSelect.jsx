// @flow
import React from 'react';

type propTypes = {
  meta: Object,
  input: Object,
}

const WithReduxFormSelect = (WrappedComponent) => {
  const Component = ({
    meta,
    input: {
      value, name, onChange, onBlur, onFocus,
    },
    ...rest
  }:propTypes) => (
    <WrappedComponent
      value={value}
      name={name}
      onChange={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
      error={meta.touched && meta.error}
      messageError={meta.error}
      {...rest}
    />
  );

  return Component;
};

export default WithReduxFormSelect;
