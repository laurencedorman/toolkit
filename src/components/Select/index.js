export { default } from './Select';
export { default as WithReduxFormSelect } from './WithReduxFormSelect';
