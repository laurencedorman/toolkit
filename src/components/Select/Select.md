Select playground:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'visa', label: 'Visa' },
    { value: 'master', label: 'MasterCard' },
    { value: 'amex', label: 'American Express' },
    { value: 'cash', label: 'Cash' },
    { value: 'other', label: 'Other' },
  ],
  value: '',
  status: 'blurred',
};

<div>
  <Select
    label="Payment method:"
    name="name"
    options={state.options}
    placeholder="---"
    className="myselect"
    onChange={event => setState({ value: event.target.value })}
    onBlur={event => setState({ status: 'blurred' })}
    onFocus={event => setState({ status: 'focused' })}
    value={state.value}
    required
  />
  <p style={{ marginTop: 24, marginBottom: 0, fontSize: 12, color: '#999' }}>
    Value: {state.value}
  </p>
  <p style={{ marginTop: 2, fontSize: 12, color: '#999' }}>
    Select is: {state.status}
  </p>
</div>;
```

Select with helper:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'apple', label: 'Apples' },
    { value: 'orange', label: 'Oranges' },
    { value: 'banana', label: 'Bananas' },
    { value: 'strawberry', label: 'Strawberries' },
  ],
  value: 'orange',
};

<Select
  label="Label"
  name="name"
  required
  options={state.options}
  helper="Helper text to display some information!"
  onChange={event => setState({ value: event.target.value })}
  value={state.value}
/>;
```

Select with error:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'beach', label: 'Beach' },
    { value: 'mountain', label: 'Mountain' },
    { value: 'jungle', label: 'Jungle' },
    { value: 'home', label: 'Home' },
  ],
  value: '',
};

<Select
  label="Vacationing at:"
  name="name"
  required
  options={state.options}
  placeholder="Choose your option"
  error="This is a required field"
  onChange={event => setState({ value: event.target.value })}
  value={state.value}
/>;
```

Select disabled:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'ocean', label: 'Ocean' },
    { value: 'mountain', label: 'Mountain' },
    { value: 'campagn', label: 'Campagn' },
  ],
  value: '',
};

<Select
  label="Label"
  name="name"
  options={state.options}
  placeholder="Choose your option"
  disabled
  onChange={event => setState({ value: event.target.value })}
  value={state.value}
/>;
```

Select with disabled option:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'ocean', label: 'Ocean' },
    { value: 'mountain', label: 'Mountain', disabled: true },
    { value: 'campagn', label: 'Campagn' },
  ],
  value: '',
};

<Select
  label="Label"
  name="name"
  options={state.options}
  placeholder="Choose your option"
  onChange={event => setState({ value: event.target.value })}
  value={state.value}
/>;
```

Select with small size:

```js
import Select from './Select';
initialState = {
  options: [
    { value: 'ocean', label: 'Ocean' },
    { value: 'mountain', label: 'Mountain' },
    { value: 'campagn', label: 'Campagn' },
  ],
  value: '',
};

<Select
  required
  label="Light Select"
  name="light"
  options={state.options}
  themeLight
  placeholder="Select with small size"
  onChange={event => setState({ value: event.target.value })}
  value={state.value}
/>;
```
