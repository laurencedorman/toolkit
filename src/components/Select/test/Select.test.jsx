/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Select from '../Select';

describe('Select', () => {
  const options = [
    { value: 'visa', label: 'Visa' },
    { value: 'master', label: 'MasterCard' },
    { value: 'amex', label: 'American Express', disabled: true },
    { value: 'cash', label: 'Cash' },
    { value: 'other', label: 'Other' },
  ];

  it('should render correctly', () => {
    const wrapper = shallow(
      <Select
        label="This is a label"
        options={options}
        name="testName"
        id="testId"
        className="test-class"
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should not have a defaultValue, it makes the component not to update correctly to immidiate value changes after first render', () => {
    const wrapper = shallow(
      <Select options={options} onChange={() => {}} value="" />
    );

    expect(wrapper.find('select').prop('defaultValue')).toBe(undefined);
  });

  it('should render required elements', () => {
    const wrapper = shallow(
      <Select
        label="This is a label"
        options={options}
        required
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper.find('.label').text()).toBe('This is a label*');
    expect(wrapper.find('select').prop('required')).toBe(true);
  });

  it('should render disabled elements', () => {
    const wrapper = shallow(
      <Select options={options} disabled onChange={() => {}} value="" />
    );

    expect(wrapper.find('.select.disabled')).toHaveLength(1);
    expect(wrapper.find('select').prop('disabled')).toBe(true);
  });

  it('should render helper correctly', () => {
    const wrapper = shallow(
      <Select
        options={options}
        helper="This is a helper"
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper.find('span.message')).toHaveLength(1);
    expect(wrapper.find('span.helperMessage').text()).toBe('This is a helper');
  });

  it('should render error correctly', () => {
    const wrapper = shallow(
      <Select
        options={options}
        error="You have an error"
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper.find('.select.hasError')).toHaveLength(1);
    expect(wrapper.find('span.message')).toHaveLength(1);
    expect(wrapper.find('span.errorMessage').text()).toBe('You have an error');
  });

  it('should render themeLight', () => {
    const wrapper = shallow(
      <Select options={options} themeLight onChange={() => {}} value="" />
    );

    expect(wrapper.find('.select.light')).toHaveLength(1);
  });

  it('should render with placeholder', () => {
    const wrapper = shallow(
      <Select
        options={options}
        placeholder="Choose"
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper.find('option')).toHaveLength(6);
    expect(wrapper.find('option:first-child').text()).toBe('Choose');
    expect(wrapper.find('option:first-child').prop('value')).toBe('');
    expect(wrapper.find('.select.noOptionSelected')).toHaveLength(1);
  });

  it('should render extra selectProps correctly', () => {
    const wrapper = shallow(
      <Select
        options={options}
        autoComplete="cc-type"
        onChange={() => {}}
        value=""
      />
    );

    expect(wrapper.find('.control').prop('autoComplete')).toEqual('cc-type');
  });

  it('should change value when new value prop is passed', () => {
    const wrapper = shallow(
      <Select
        label="This is a label"
        options={options}
        required
        onChange={newValue => {
          wrapper.setProps({ value: newValue });
        }}
        value=""
      />
    );

    expect(wrapper.find('select').prop('value')).toBe('');
    wrapper.find('select').simulate('change', 'New Test Value');
    expect(wrapper.find('select').prop('value')).toBe('New Test Value');
  });
});
