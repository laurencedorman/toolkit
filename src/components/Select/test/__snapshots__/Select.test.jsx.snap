// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
<div
  className="toolkit select test-class"
  data-qa=""
>
  <label
    htmlFor="testId"
  >
    <span
      className="label"
    >
      This is a label
    </span>
    <span
      className="mask"
    >
      <select
        aria-required={false}
        className="control"
        disabled={false}
        id="testId"
        name="testName"
        onChange={[Function]}
        required={false}
        value=""
      >
        <option
          key="0-visa"
          value="visa"
        >
          Visa
        </option>
        <option
          key="1-master"
          value="master"
        >
          MasterCard
        </option>
        <option
          disabled={true}
          key="2-amex"
          value="amex"
        >
          American Express
        </option>
        <option
          key="3-cash"
          value="cash"
        >
          Cash
        </option>
        <option
          key="4-other"
          value="other"
        >
          Other
        </option>
      </select>
      <span
        className="arrow"
      />
    </span>
  </label>
</div>
`;
